Add LDAP authentication with Elytron LDAP realm

hartimcwildfly opened the pull request:
currently tests and documentation are missing. Just tested it in my dev stage -> works
Some tips how to mock the ldap server in unit tests would be nice :)

gsmet reviewed (changes requested):
Thanks. Just a marker that we will need all this to be squashed.

hartimcwildfly commented:
Ready for merging. I hope everything is fine :)

gsmet commented:
@hartimcwildfly could you squash everything into a single commit and rebase? Thanks!

hartimcwildfly commented:
Rebased and squashed everything (Sorry I did not understand what you meant with squashing in your first comment). Let's see if everything still works.

hartimcwildfly commented:
@gsmet Could you take a look at the pipelines? Seems like your CI is broken:
##[error]No agent found in pool Azure Pipelines which satisfies the specified demands:
     Agent.Version -gtVersion 2.161.0

hartimcwildfly commented:
All checks passed. Ready for review/merge.

gsmet commented on docs/src/main/asciidoc/security-ldap.adoc:
```diff
@@ -0,0 +1,226 @@
+
+include::./attributes.adoc[]
+
+This guide demonstrates how your Quarkus application can use an ldap server to authenticate and authorize your user identities.
```
```suggestion
This guide demonstrates how your Quarkus application can use an LDAP server to authenticate and authorize your user identities.
```

gsmet commented on docs/src/main/asciidoc/security-ldap.adoc:
```diff
@@ -0,0 +1,226 @@
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+= Quarkus - Using Security with a LDAP Realm
```
```suggestion
= Quarkus - Using Security with an LDAP Realm
```

gsmet commented on docs/src/main/asciidoc/security-ldap.adoc:
```diff
@@ -0,0 +1,226 @@
+----
+
+This command generates a Maven project, importing the `elytron-security-ldap` extension
+which is an https://docs.wildfly.org/17/WildFly_Elytron_Security.html#ldap-security-realm[`wildfly-elytron-realm-ldap`] adapter for Quarkus applications.
```
```suggestion
which is a https://docs.wildfly.org/17/WildFly_Elytron_Security.html#ldap-security-realm[`wildfly-elytron-realm-ldap`] adapter for Quarkus applications.
```

hartimcwildfly replied:
"an [...] adapter" is correct.

gsmet commented on docs/src/main/asciidoc/security-ldap.adoc:
```diff
@@ -0,0 +1,226 @@
+quarkus.security.ldap.identity-mapping.attribute-mapping.filter-base-dn=ou=roles,ou=tool,o=YourCompany,c=DE
+--
+
+The `elytron-security-ldap` extension requires at least one principal query to autenticate the user and its identity.
```
```suggestion
The `elytron-security-ldap` extension requires at least one principal query to authenticate the user and its identity.
```

hartimcwildfly replied:
This is semantically wrong. I'm going to fix this, too. (PrincipalQueries are from JDBC Realm)

gsmet commented on docs/src/main/asciidoc/security-ldap.adoc:
```diff
@@ -0,0 +1,226 @@
+quarkus.security.ldap.identity-mapping.attribute-mapping.from=cn
+quarkus.security.ldap.identity-mapping.attribute-mapping.to=groups
+quarkus.security.ldap.identity-mapping.attribute-mapping.filter=(member=uid={0})
+quarkus.security.ldap.identity-mapping.attribute-mapping.filter-base-dn=ou=roles,ou=tool,o=YourCompany,c=DE
```
I'm surprised we can have only one `attribute-mapping`?

hartimcwildfly replied:
Hmm you are right, one should be able to define multiple attribute-mappings. But currently the quarkus config only allows Map<String, AttributeMappingConfig> but not Map<Integer, AttributeMappingConfig> or List<AttributeMappingConfig>. For my usecase a single attribute-mapping was enough.

hartimcwildfly replied:
Added possibility to define multiple attribute-mappings. Pushing soon.

gsmet commented on docs/src/main/asciidoc/security-ldap.adoc:
```diff
@@ -0,0 +1,226 @@
+
+== Testing the Application
+
+The application is now protected and the identities are provided by our ldap server.
```
```suggestion
The application is now protected and the identities are provided by our LDAP server.
```

gsmet commented on extensions/elytron-security-ldap/deployment/src/test/java/io/quarkus/elytron/security/ldap/LdapSecurityRealmTest.java:
```diff
@@ -0,0 +1,178 @@
+        inMemoryDirectoryServerConfig.addAdditionalBindCredentials("uid=admin,ou=system", "secret");
+        ldapServer = new InMemoryDirectoryServer(inMemoryDirectoryServerConfig);
+        ldapServer.importFromLDIF(true, new LDIFReader(ClassLoader.getSystemResourceAsStream("quarkus-io.ldif")));
+        ldapServer.startListening();
```
I wonder if this should be implemented as a test resource instead? See what we did for H2 for instance (`H2DatabaseTestResource`). For now, the test resource could be in this artifact, no need to create a test artifact.

hartimcwildfly replied:
I did this for the integration tests, but I did not test it with the unit tests.
I'm going to take another look and compact it. The ITs and unit tests are similar but I oriented on the elytron jdbc extension which also has overlapping tests.

hartimcwildfly replied:
Outsourced this in an extra TestResource. Pushing soon.

gsmet commented on extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/config/AttributeMappingConfig.java:
```diff
@@ -0,0 +1,35 @@
+    /**
+     * The roleAttributeId from which is mapped (e.g. "cn")
+     */
+    @ConfigItem(name = "from")
```
You can remove all the names if the property has the right name.

hartimcwildfly replied:
Ok, I'm going to fix this.

gsmet commented on extensions/elytron-security-ldap/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,10 @@
+  guide: "https://quarkus.io/guides/security-ldap"
+  categories:
+  - "security"
+  status: "stable"
```
Maybe make it `preview` for some time?

hartimcwildfly replied:
ok

gsmet reviewed (dismissed):
I added a few questions and suggestions.

I'm not the security expert here so we will need an approval from @sberyozkin too.

hartimcwildfly commented:
Ready for another review. Should I squash it again?

hartimcwildfly commented:
Deleted a duplicate file and squashed everything again.

hartimcwildfly commented:
Deleted a unused maven dependency. Now ready for merge.
@sberyozkin and @gsmet  please review.

loicmathieu commented on extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java:
```diff
@@ -0,0 +1,82 @@
+            public Provider[] get() {
+                return new Provider[] { new WildFlyElytronProvider() };
+            }
+        };
```
This will trigger the creation of a `WildFlyElytronProvider` object on each requests.

Please use a constant to hold the provider.

See PR #6358 that fixes the same issue on the existing security extensions.

hartimcwildfly replied:
Fixed.

loicmathieu reviewed (changes requested):
I only review this PR for a known performance issue, please fix the performance issue as requested.

loicmathieu commented on extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java:
```diff
@@ -0,0 +1,79 @@
+     * @return - runtime value wrapper for the SecurityRealm
+     */
+    public RuntimeValue<SecurityRealm> createRealm(LdapSecurityRealmConfig config) {
+        Supplier<Provider[]> providers = () -> PROVIDERS;
```
We try to avoid lambda inside Quarkus extension code

hartimcwildfly replied:
Reverted to anonymous function instead of lambda.

loicmathieu commented:
@hartimcwildfly thanks for the performance fix I asked, it's OK for me. I added a small comment on the usage of lambda.

loicmathieu reviewed (approved):
LGTM, but only for the perf issue I mention, don't count me as a reviewer.

gsmet commented:
Current status is: we are waiting for @sberyozkin to review that one.

sberyozkin commented:
@hartimcwildfly @gsmet sorry for a delay, I was away. I'm catching up and will try to comment soon, thanks

sberyozkin commented:
@hartimcwildfly @gsmet While I'm not an LDAP expert, this PR does look quite perfect to me, as it is neatly done and simply to follow.  Let me approve. Hi Stuart @stuartwdouglas, have a look please too if you can find some time, I'd be happy to merge tomorrow my time.

stuartwdouglas commented on bom/runtime/pom.xml:
```diff
@@ -176,6 +176,7 @@
         <logstash-gelf.version>1.13.0</logstash-gelf.version>
         <jsch.version>0.1.55</jsch.version>
         <jzlib.version>1.1.1</jzlib.version>
+        <unboundid-ldap.version>4.0.13</unboundid-ldap.version>
```
It looks like this is only used for testing? If so it should be in the build parent rather than the bom.

hartimcwildfly replied:
Makes sense. Fixed.

Commit:
added elytron ldap authentication
